### PR TITLE
[SPARK-49246][SQL][TESTS][FOLLOWUP] Fix flaky CollationSuite

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr is following up https://github.com/apache/spark/pull/48019.

### Why are the changes needed?
Fix flaky `CollationSuite `.
After this PR, GA's `CollationSuite` failed, let's fix it first.
https://github.com/panbingkun/spark/actions/runs/10765984604/job/29851047311
image

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Update existed UT.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by MaxGekk — COMMENTED** [review]
cc @amaliujia @cloud-fan Please, review this since your changes cause the test failure.

**Comment by cloud-fan** [comment]
I'm confused, the test were all green when I merged it.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 1]
```diff
@@ -947,7 +947,10 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
     withTable(table) {
       sql(s"create table $table (a array<string collate utf8_lcase>) using parquet")
       sql(s"insert into $table values (array('aaa')), (array('AAA'))")
-      checkAnswer(sql(s"select distinct a from $table"), Seq(Row(Seq("aaa"))))
+      val result = sql(s"select distinct a from $table").collect()
+      assert(result.length === 1)
+      val data = result.head.get(0).asInstanceOf[scala.collection.mutable.ArraySeq[String]]
+      assert(data === Array("aaa") || data === Array("AAA"))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:953]: so the test becomes flaky? Reading the test query I think both `aaa` and `AAA` are correct result. cc @stefankandic

**Comment by MaxGekk** [comment]
I guess we should extend `checkAnswer`, so, it should be aware of collations in comparing of resulted strings. WDYT?

**Comment by panbingkun** [comment]
> I guess we should extend `checkAnswer`, so, it should be aware of collations in comparing of resulted strings. WDYT?

I want to do this, but maybe we can let it in first and turn GA green.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@panbingkun `branch-3.5` is flaky too?

**Comment by panbingkun** [comment]
> @panbingkun `branch-3.5` is flaky too? Should we backport your changes to `branch-3.5`?

Branch-3.5 does not have this class, as follows:
(base) ➜  spark-community git:(branch-3.5) ✗ find . -name "CollationSuite.scala"
(base) ➜  spark-community git:(branch-3.5) ✗

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 1 above)
Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:953]: Yes, it becomes flaky. The returned value is not non-deterministic.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 2]
```diff
@@ -947,7 +947,10 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
     withTable(table) {
       sql(s"create table $table (a array<string collate utf8_lcase>) using parquet")
       sql(s"insert into $table values (array('aaa')), (array('AAA'))")
-      checkAnswer(sql(s"select distinct a from $table"), Seq(Row(Seq("aaa"))))
+      val result = sql(s"select distinct a from $table").collect()
+      assert(result.length === 1)
+      val data = result.head.get(0).asInstanceOf[scala.collection.mutable.ArraySeq[String]]
```
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:952]:
```suggestion
      val data = result.head.getSeq[String](0)
```

Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala]: Updated, thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:953]: You mean just non-deterministic, I guess.

**Comment by zhengruifeng** [comment]
thanks @panbingkun for working on it, I was also confused for a while

**Comment by cloud-fan** [comment]
We should discuss the general fix carefully. I don't think we should do case insensitive string comparison in `checkAnswer` blindly. For example, `aAa` is not a correct answer while `aaa` and `AAA` are.

**Review by cloud-fan — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**